File: handy-js/quiz/q16_best_time_to_buy_and_sell_stock_easy.js
/**
 * Q16) Best Time to Buy and Sell Stock
 *
 * AI-BEST:
 * Track min price so far and max profit so far.
 * Time: O(n), Space: O(1)
 */

function maxProfitBest(prices) {
  // Step 1: Track lowest buy price seen so far.
  // Why: best sell profit at day i is always relative to the cheapest earlier day.
  let minPrice = Infinity;
  let bestProfit = 0;

  // Step 2: Scan prices once and update best profit.
  for (const price of prices) {
    minPrice = Math.min(minPrice, price);
    bestProfit = Math.max(bestProfit, price - minPrice);
  }

  // Step 3: Return max one-transaction profit.
  return bestProfit;
}

function runDemo() {
  console.log("Q16: Best Time to Buy and Sell Stock");
  const cases = [
    [7, 1, 5, 3, 6, 4],
    [7, 6, 4, 3, 1],
    [2, 4, 1],
  ];

  for (const prices of cases) {
    console.log(`Input: prices=${JSON.stringify(prices)}`);
    console.log("  BEST:", maxProfitBest(prices));
  }
}

runDemo();

File: handy-java/quiz/q16_best_time_to_buy_and_sell_stock_easy.java
import java.util.Arrays;

/**
 * Q16) Best Time to Buy and Sell Stock
 *
 * AI-BEST:
 * Track lowest price so far and best profit.
 * Time: O(n), Space: O(1)
 */
public class q16_best_time_to_buy_and_sell_stock_easy {
    static int maxProfitBest(int[] prices) {
        // Step 1: Track minimum buy price while scanning.
        // Why: best profit on any day depends on cheapest prior day.
        int minPrice = Integer.MAX_VALUE;
        int bestProfit = 0;

        // Step 2: Update min and best profit in one pass.
        for (int price : prices) {
            minPrice = Math.min(minPrice, price);
            bestProfit = Math.max(bestProfit, price - minPrice);
        }

        // Step 3: Return maximum one-transaction profit.
        return bestProfit;
    }

    public static void main(String[] args) {
        System.out.println("Q16: Best Time to Buy and Sell Stock");
        int[][] cases = {
                {7, 1, 5, 3, 6, 4},
                {7, 6, 4, 3, 1},
                {2, 4, 1}
        };

        for (int[] prices : cases) {
            System.out.println("Input: prices=" + Arrays.toString(prices));
            System.out.println("  BEST: " + maxProfitBest(prices));
        }
    }
}
